### PR TITLE
Added debug for waypoints on wanderingbots

### PR DESCRIPTION
**Pull request by Conan1996-2**
## Changes Proposed:

This PR proposes changes to:
-  [x ] Core (units, players, creatures, game systems).

## SOURCE:

The changes have been validated through:
- [x ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

[https://github.com/trickerer/AzerothCore-wotlk-with-NPCBots/compare/npcbots_3.3.5...Conan1996-2:AzerothCore-wotlk-with-NPCBots:npcbots_3.3.5](https://github.com/trickerer/AzerothCore-wotlk-with-NPCBots/compare/npcbots_3.3.5...Conan1996-2:AzerothCore-wotlk-with-NPCBots:npcbots_3.3.5)

## Tests Performed:

This PR has been:
- [ x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

Right click on a wandering bot and select the "Debug" then "<Waypoint Data>"
Information shown will be:

1. current waypoint id, x, y, z, proximity
2. Next waypoint id, x, y, z, proximity
3. current x, y, z and distance from waypoint.

This will be useful information if there are bots "Hearthstoning", falling through maps or other bot errors.

*EDIT*
Waypoint flag (BOTWP_FLAG_NOT_A_START_POINT) has been added and can be entered in the creature_template_npcbot_wander_nodes flags column for waypoints you do not want bots teleporting to when bots need to find a new waypoint. 

*EDIT* Removed the following for further local testing
<strike>
Also added two functions, main "HandleNpcBotOrderAttackCommand" and helper "FindNearbyTargetGuidByMask"

use:
`.npcbot order attack`
OR create a macro that will set the raid target on the target and order the attack. Make sure to change the number 2 to the correct icon number
```
/script SetRaidTarget("target", 2);
.npcbot order attack

```
The purpose:
1. Allow tanking multiple targets with multiple tanks/offtanks using symbols from "TankTargetIconMask" and "OffTankTargetIconMask" in the .conf file at the same time with one command. When the command is given, each tank will individually attack a separate bot (split pull) with a raid target icon predefined in your .conf file.
2. Easier to mark a target and attack in dungeon/raid than using the .npcbot order pull "botname" "icon" given you have to change the name each time you get a new tank bot.
3. If playing a healer, you don't have to run in and gain "Agro" first, stand back and just use the command attack.

Requirements:
1) Must have at least one bot tank in the group/raid
2) Must have the intended valid target marked with an appropriate icon from the .conf file.
</strike>

- [x ] This pull request can be tested by following the reproduction steps provided in the linked issue

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/AI/NpcBots/bot_ai.cpp [hunk 1]
```diff
@@ -11401,6 +11401,19 @@ bool bot_ai::OnGossipSelect(Player* player, Creature* creature/* == me*/, uint32
                     }
                     break;
                 }
+                case 7: // Show Waypoint Data
+                {
+                    ChatHandler ch(player->GetSession());
+                    if (_travel_node_last && _travel_node_cur) {
```
Comment (trickerer) [inline src/server/game/AI/NpcBots/bot_ai.cpp]: Please fix the codestyle:
- Do not use trailing opening curly bracket `if () {`
- Same with trailing `} else` at 11412

**Comment by trickerer** [comment]
The resulting behavior (as I see it) is no different than from using `.npcbot order pull` command. You use BOT_ORDER_PULL order type, this means pulling bot is going to attack the target once and then run back (if follow distance is not large enough and tanking target icon != bot's priority target icon).
Setting priority target for all bots in a command is not a good idea and I don't expect you to be able to implement a new order type (requires extra checks in `GetTargets()` and `GlobalUpdate()`)
So why is this not a modification of existing pull command? For example you could make this functionality fire in case no arguments were provided.
The point I'm trying to make here is this behavior doesn't necessarily match command name

**Comment by Conan1996-2** [comment]
You are correct, that is why I said I was expanding on your idea in the original chat, having to 1) Not know the name of your tank(s) and/or Offtank(s), and 2) Being able to send more then one unit in the command to attack, split pull, etc. I original started to expand your function but it was becoming to messy, so I just created a new one. Also, I want to expand this by just using the /attacktarget like you use the /wave emotes. (Hence why I used the attack command, not called it charge, or whatever)

I have tested this with a tank and 2 off tanks and It works great, all three go to their respective targets and attack and keep agro.

I have changed the code: It now attacks and stays with the target until dead. I need to work on the icons a little bit.

**Comment by Conan1996-2** [comment]
I have changed the code. I initially used your code on the pull command to pull the target(s). I have since changed the code, now the tank(s) and offtank(s) attack regardless of distance and stay and fight.

                    bot->GetMotionMaster()->Clear();                      
                    bot->GetMotionMaster()->MoveChase(tank_target, 0.0f, 0.0f);                   
                    ai->AttackStart(tank_target);                   
                    bot->SetInCombatWith(tank_target);                    
                    tank_target->SetInCombatWith(bot);

**Comment by trickerer** [comment]
> I have changed the code. I initially used your code on the pull command to pull the target(s). I have since changed the code, now the tank(s) and offtank(s) attack regardless of distance and stay and fight.
> 
> ```
>                 bot->GetMotionMaster()->Clear();                      
>                 bot->GetMotionMaster()->MoveChase(tank_target, 0.0f, 0.0f);                   
>                 ai->AttackStart(tank_target);                   
>                 bot->SetInCombatWith(tank_target);                    
>                 tank_target->SetInCombatWith(bot);
> ```

Sorry but this is a no-go: we can't use `SetInCombatWith()` in AI or commands - that's an exploit which allows players to put any unit in combat at will. This function is only supposed to be called by core internals and scripts (events), search for existing usages.

**Comment by Conan1996-2** [comment]
> > I have changed the code. I initially used your code on the pull command to pull the target(s). I have since changed the code, now the tank(s) and offtank(s) attack regardless of distance and stay and fight.
> > ```
> >                 bot->GetMotionMaster()->Clear();                      
> >                 bot->GetMotionMaster()->MoveChase(tank_target, 0.0f, 0.0f);                   
> >                 ai->AttackStart(tank_target);                   
> >                 bot->SetInCombatWith(tank_target);                    
> >                 tank_target->SetInCombatWith(bot);
> > ```
> 
> Sorry but this is a no-go: we can't use `SetInCombatWith()` in AI or commands - that's an exploit which allows players to put any unit in combat at will. This function is only supposed to be called by core internals and scripts (events), search for existing usages.

Understood... Ill have to dig through the azerothcore code and see how hunters "Pets" are commanded to attack a target, even when out of range.

**Comment by Conan1996-2** [comment]
Ok, I removed the "Attack" until further coding and testing can be done so this pull can go through unless there are issues.

This pull does three things, It corrects the "waypoint" proximity, adds waypoint debug for GMS and it adds a flag so a waypoint will NOT be used during a search for a waypoint if it needs to teleport to one.

**Review by trickerer — CHANGES_REQUESTED** [review]

**Comment by Conan1996-2** [comment]
All set, old coding habit

**Comment by trickerer** [comment]
Ok, but please remove trailing whitespaces - even in gh code viewer I see at least 4 places. Don't make me fix that for you after the merge

**Comment by Conan1996-2** [comment]
> Ok, but please remove trailing whitespaces - even in gh code viewer I see at least 4 places. Don't make me fix that for you after the merge

Ok, I'm using the web github, it does not show trailing whitespace. So I went line by line and looked for trailing whitespaces.

**Review by trickerer — APPROVED** [review]
